doc/SAI-Proposal-ICMP-ECHO.md

```diff
@@ -0,0 +1,640 @@
+
+When control plane application carries out the link probe (software link failure detection), the system operates at a slow, periodic exchange of probe requests and replies since it uses the kernel network stack to send and receive control packets. This may prolong link failure detection. To decrease this detection interval, the application could utilize the HW/ASIC to send and receive link probe packets at a quicker rate.
+
+The objective of this document is to propose a SAI specification to simplify the ICMP-based probe as a basic service. The service primitives offered by ICMP echo include creating, destroying, and modifying a session, considering the destination address and other parameters. In response, ICMP Echo sends a signal to its clients when the ICMP Echo session starts or stops.
+
+This specification suggests:
+- The idea of an ICMP echo object or session that can be transferred to HW/ASIC or other modules for faster link detection.
+- The introduction of SAI APIs to outline ICMP echo session properties.
+- The introduction of a switch-level notification API to monitor changes in ICMP echo session states.
+
+
+## 1.1.0 Functional requirements of ICMP echo session
+- Transmit ICMP echo requests (probes) at regular, configurable intervals.
+- Identify and report changes in probe state to the control plane application.
+- Enable applications to set customized transmit and receive intervals.
+- Allow configuration of IP encapsulation parameters.
+- Establish a unique identification for each probing session.
+- Determine whether an offloaded session should process ICMP probing packets.
+- Offer statistics on ICMP echo probe and reply activities.
+
+## 2.0 Protocol overview and ICMP echo session
+ICMP echo is simple network link connectivity test protocol, similar to that most of the L3 routing protocols. ICMP Echo session runs between pair of devices exchanging ICMP echo request and reply packets.
+
+### 2.1.0 ICMP echo request and reply packet ###
+
+```C
+    0                   1                   2                   3
+    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |     Type      |     Code      |          Checksum             |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |           Identifier          |        Sequence Number        |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |     Data ...
+   +-+-+-+-+-
+```
+
+### 2.2.0 ICMP echo payload packet ###
+
+Besides the ICMP echo request and reply packets, additional information about the ICMP echo session is essential and is transmitted within the payload.
+
+***Session ID:*** To correlate probe and reply packets with a session, especially when the session is offloaded to hardware, a unique identifier is required. This distinct identifier is conveyed within the ICMP payload field.
```

Comment:
No RFC. It is a MSFT proprietary format. its definition can be found in https://github.com/sonic-net/SONiC/blob/master/doc/dualtor/active_active_hld.md (section 3.3.1 Link Prober) 
The session ID is required because in the dual tor deployment virtual IPs are used. Ethernet source and destination addresses as well as the IP source and destination address are the same for packets originating from both ToR devices.

Reply:
Thanks for the reference.  Please clarify the below differences between the formats defined in the Dual TOR HLD and this SAI proposal.

1.  Sequence number is 8B in the SONiC HLD, whereas it is 4B in the SAI PR.
2.  SONiC HLD has a 8B GUID, SAI PR has a 8B Session-ID. Are these two different names for the same field?

Reply:
1) sequence number is 8B. fixed the doc. 
2) 8B Session-ID is same as 8B GUID of SONiC HLD. I fixed the doc by changing the name.